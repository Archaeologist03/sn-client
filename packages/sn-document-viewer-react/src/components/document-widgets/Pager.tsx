import IconButton from '@material-ui/core/IconButton'
import TextField from '@material-ui/core/TextField'
import FirstPage from '@material-ui/icons/FirstPage'
import LastPage from '@material-ui/icons/LastPage'
import NavigateBefore from '@material-ui/icons/NavigateBefore'
import NavigateNext from '@material-ui/icons/NavigateNext'
import React, { useCallback, useEffect, useState } from 'react'
import { useViewerState } from '../../hooks/useViewerState'
import { useDocumentData, useLocalization } from '../../hooks'

/**
 * Defines the own props for the PagerState component
 */
export interface PagerState {
  currentPage: number
  lastPage: number
}

export const PagerWidget: React.FC = () => {
  const viewerState = useViewerState()
  const localization = useLocalization()
  const { documentData } = useDocumentData()

  const [currentPage, setCurrentPage] = useState(viewerState.activePages[0] || 1)
  const [lastPage, setLastPage] = useState(documentData.pageCount || 1)

  useEffect(() => {
    setCurrentPage(viewerState.activePages[0] || 1)
  }, [viewerState.activePages])

  useEffect(() => {
    setLastPage(documentData.pageCount || 1)
  }, [documentData.pageCount])

  const gotoPage = useCallback(
    (pageNo: number) => {
      viewerState.updateState({ activePages: [pageNo] })
    },
    [viewerState],
  )

  return (
    <div style={{ display: 'inline-block' }}>
      <IconButton disabled={currentPage <= 1} title={localization.firstPage} onClick={() => gotoPage(1)} id="FirstPage">
        <FirstPage />
      </IconButton>

      <IconButton
        disabled={currentPage <= 1}
        title={localization.previousPage}
        onClick={() => gotoPage(viewerState.activePages[0] - 1)}
        id="NavigateBefore">
        <NavigateBefore />
      </IconButton>

      <TextField
        title={localization.gotoPage}
        value={currentPage}
        onChange={ev => gotoPage(parseInt(ev.currentTarget.value, 10))}
        type="number"
        required={true}
        InputLabelProps={{
          shrink: true,
        }}
        inputProps={{ min: 1, max: lastPage }}
        margin="dense"
      />

      <IconButton
        disabled={currentPage >= lastPage}
        title={localization.nextPage}
        color={'primary'}
        onClick={() => gotoPage(viewerState.activePages[0] + 1)}
        id="NavigateNext">
        <NavigateNext />
      </IconButton>

      <IconButton
        disabled={currentPage >= lastPage}
        title={localization.lastPage}
        onClick={() => gotoPage(lastPage)}
        id="LastPage">
        <LastPage />
      </IconButton>
    </div>
  )
}
