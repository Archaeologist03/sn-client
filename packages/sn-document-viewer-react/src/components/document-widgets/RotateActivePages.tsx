import IconButton from '@material-ui/core/IconButton'
import RotateLeft from '@material-ui/icons/RotateLeft'
import RotateRight from '@material-ui/icons/RotateRight'

import React, { useCallback } from 'react'
import { PreviewImageData } from '../../models'
import { RootReducerType, rotateImages } from '../../store'
import { useLocalization, useViewerState } from '../../hooks'

/**
 * maps state fields from the store to component props
 * @param state the redux state
 */
export const mapStateToProps = (state: RootReducerType) => {
  return {
    pages: state.sensenetDocumentViewer.previewImages.AvailableImages as PreviewImageData[],
    activePages: state.sensenetDocumentViewer.viewer.activePages,
  }
}

/**
 * maps state actions from the store to component props
 * @param state the redux state
 */
export const mapDispatchToProps = {
  rotateImages,
}

/**
 * Component that allows active page rotation
 */
export const RotateActivePagesComponent: React.FC = () => {
  const localization = useLocalization()
  const viewerState = useViewerState()

  const rotateDocumentLeft = useCallback(() => {
    // ToDo
    // props.rotateImages(viewerState.activePages, -90)
  }, [])

  const rotateDocumentRight = useCallback(() => {
    // ToDo
    // props.rotateImages(viewerState.activePages, 90)
  }, [])

  /**
   * renders the component
   */
  return (
    <div style={{ display: 'inline-block' }}>
      <IconButton
        color="inherit"
        title={localization.rotateDocumentLeft}
        onClick={() => rotateDocumentLeft()}
        id="RotateActiveLeft">
        <RotateLeft />
      </IconButton>
      <IconButton
        color="inherit"
        title={localization.rotateDocumentRight}
        onClick={() => rotateDocumentRight()}
        id="RotateActiveRight">
        <RotateRight />
      </IconButton>
    </div>
  )
}
