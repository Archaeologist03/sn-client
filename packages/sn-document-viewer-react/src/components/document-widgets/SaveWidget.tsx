import IconButton from '@material-ui/core/IconButton'
import Save from '@material-ui/icons/Save'
import React, { useCallback } from 'react'
import { saveChanges } from '../../store'
import { useDocumentData, useDocumentPermissions, useDocumentViewerApi } from '../../hooks'

/**
 * Document widget component for saving document state
 */
export const SaveDocument: React.FC = () => {
  const api = useDocumentViewerApi()
  const document = useDocumentData()
  const permissions = useDocumentPermissions()

  const save = useCallback(() => {
    permissions.canEdit && api.saveChanges(document, this.props.pages)
  }, [api, document, permissions.canEdit])

  return (
    <div style={{ display: 'inline-block' }}>
      <IconButton
        disabled={!this.props.hasChanges || !permissions.canEdit}
        title={saveChanges}
        onClick={save}
        id="Save">
        <Save />
      </IconButton>
    </div>
  )
}
