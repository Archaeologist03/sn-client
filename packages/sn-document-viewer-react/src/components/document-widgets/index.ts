export { DocumentTitlePager } from './DocumentTitlePager'
export { Download } from './DownloadWidget'
export { Print } from './PrintWidget'
export { Share } from './ShareWidget'
export { RotateActivePagesWidget } from './RotateActivePages'
export { RotateDocumentWidget } from './RotateDocument'
export { SaveWidget } from './SaveWidget'
export { SearchBar } from './SearchBar'
export { ToggleRedactionWidget } from './ToggleRedaction'
export { ToggleBase } from './ToggleBase'
export { ToggleShapesWidget } from './ToggleShapes'
export { ToggleThumbnailsWidget } from './ToggleThumbnails'
export { ToggleWatermarkWidget } from './ToggleWatermark'
export { ToggleCommentsWidget } from './ToggleComments'
export { ZoomInOutWidget } from './ZoomInOut'
export { ZoomModeWidget } from './ZoomMode'
