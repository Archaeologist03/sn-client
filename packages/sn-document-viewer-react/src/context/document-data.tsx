import React, { useCallback, useEffect, useState } from 'react'
import { DeepPartial, sleepAsync } from '@sensenet/client-utils'
import { DocumentData } from '../models'
import { PreviewState } from '../Enums'
import { useDocumentViewerApi, useViewerSettings } from '../hooks'

const defaultDocumentData: DocumentData & { updateDocumentData: (newData: DeepPartial<DocumentData>) => void } = {
  documentName: '',
  documentType: '',
  fileSizekB: 0,
  hostName: '',
  idOrPath: 0,
  pageAttributes: [],
  pageCount: PreviewState.Loading,
  shapes: {
    annotations: [],
    highlights: [],
    redactions: [],
  },
  error: undefined,
  updateDocumentData: () => undefined,
}

export const DocumentDataContext = React.createContext<typeof defaultDocumentData>(defaultDocumentData)

export const DocumentDataProvider: React.FC = ({ children }) => {
  const api = useDocumentViewerApi()
  const doc = useViewerSettings()

  const [docData, setDocData] = useState<DocumentData>(defaultDocumentData)

  useEffect(() => {
    const ac = new AbortController()
    ;(async () => {
      try {
        setDocData(defaultDocumentData)
        while (docData.pageCount === PreviewState.Loading && !ac.signal.aborted) {
          const result = await api.getDocumentData({
            hostName: doc.hostName,
            idOrPath: doc.documentIdOrPath,
            version: doc.version,
            abortController: ac,
          })
          setDocData(result)
          if (result.pageCount === PreviewState.Loading) {
            await sleepAsync(4000)
          }
        }
      } catch (error) {
        /** */
        if (!ac.signal.aborted) {
          setDocData({ ...defaultDocumentData, pageCount: PreviewState.ClientFailure, error: error.toString() })
          throw error
        }
      }
    })()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [api, doc.documentIdOrPath, doc.hostName, doc.version])

  const updateDocumentData = useCallback(
    (newDocData: Partial<DocumentData>) => {
      setDocData({ ...docData, ...newDocData })
    },
    [docData],
  )

  return (
    <DocumentDataContext.Provider value={{ ...docData, updateDocumentData }}>{children}</DocumentDataContext.Provider>
  )
}
