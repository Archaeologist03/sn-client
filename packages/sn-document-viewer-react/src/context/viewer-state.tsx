import React, { useCallback, useEffect, useState } from 'react'
import { deepMerge, DeepPartial } from '@sensenet/client-utils'
import { ViewerState } from '../models/viewer-state'

export const defaultViewerState: ViewerState & { updateState: (newState: DeepPartial<ViewerState>) => void } = {
  activePages: [1],
  zoomMode: 'fit',
  customZoomLevel: 3,
  showWatermark: false,
  showRedaction: true,
  showShapes: true,
  showThumbnails: false,
  fitRelativeZoomLevel: 0,
  showComments: false,
  updateState: () => {
    /** */
  },
}
export const ViewerStateContext = React.createContext(defaultViewerState)

export const ViewerStateProvider: React.FC<{ options?: Partial<ViewerState> }> = props => {
  const [state, setState] = useState<ViewerState>(deepMerge({ ...defaultViewerState }, props.options))

  useEffect(() => {
    setState(deepMerge({ ...defaultViewerState }, props.options))
  }, [props.options])

  const updateState = useCallback((newState: DeepPartial<ViewerState>) => {
    setState(deepMerge({ ...state }, newState))
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return <ViewerStateContext.Provider value={{ ...state, updateState }}>{props.children}</ViewerStateContext.Provider>
}
