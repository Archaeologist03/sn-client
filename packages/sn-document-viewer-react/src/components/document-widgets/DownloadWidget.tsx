import IconButton from '@material-ui/core/IconButton'
import CloudDownload from '@material-ui/icons/CloudDownload'
import React, { useContext } from 'react'
import { DocumentData } from '../../models'
import { LocalizationContext } from '../../context/localization-context'
import { DocumentDataContext } from '../../context/document-data'

/**
 * Own properties for the Share component
 */
export interface DownloadProps {
  download: (document: DocumentData) => void
}

/**
 * Component that allows active page rotation
 */
export const Download: React.FC<DownloadProps> = props => {
  const localization = useContext(LocalizationContext)
  const document = useContext(DocumentDataContext)
  return (
    <div style={{ display: 'inline-block' }}>
      <IconButton
        color="inherit"
        title={localization.download}
        onClick={() => props.download(document)}
        id="CloudDownload">
        <CloudDownload />
      </IconButton>
    </div>
  )
}
