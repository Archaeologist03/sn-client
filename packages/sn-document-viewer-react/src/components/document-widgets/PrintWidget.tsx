import IconButton from '@material-ui/core/IconButton'
import Print from '@material-ui/icons/Print'
import React, { useContext } from 'react'
import { DocumentData } from '../../models'
import { LocalizationContext } from '../../context/localization-context'
import { DocumentDataContext } from '../../context/document-data'

/**
 * Own properties for the Print component
 */
export interface PrintProps {
  print: (document: DocumentData) => void
}

/**
 * Component that allows active page rotation
 */
export const PrintComponent: React.FC<PrintProps> = props => {
  const localization = useContext(LocalizationContext)
  const document = useContext(DocumentDataContext)
  return (
    <div style={{ display: 'inline-block' }}>
      <IconButton color="inherit" title={localization.print} onClick={() => props.print(document)} id="Print">
        <Print />
      </IconButton>
    </div>
  )
}
