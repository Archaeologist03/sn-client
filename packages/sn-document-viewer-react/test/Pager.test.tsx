import TextField from '@material-ui/core/TextField/TextField'
import { sleepAsync } from '@sensenet/client-utils'
import { mount, shallow } from 'enzyme'
import React from 'react'
import { PagerWidget } from '../src/components/document-widgets/Pager'
import { DocumentDataContext } from '../src/context/document-data'
import { defaultViewerState, ViewerStateContext } from '../src/context/viewer-state'
import { exampleDocumentData } from './__Mocks__/viewercontext'

describe('PagerWidget component', () => {
  it('Should render without crashing', () => {
    const wrapper = shallow(<PagerWidget />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Numeric input should update the active page', async () => {
    const updateState = jest.fn()
    const wrapper = mount(
      <ViewerStateContext.Provider value={{ ...defaultViewerState, updateState, activePages: [5] }}>
        <DocumentDataContext.Provider
          value={{ documentData: { ...exampleDocumentData, pageCount: 10 }, updateDocumentData: jest.fn() }}>
          <PagerWidget />
        </DocumentDataContext.Provider>
      </ViewerStateContext.Provider>,
    )
    const changeValue = (value: number | string) =>
      (wrapper.find(TextField).prop('onChange') as any)({ currentTarget: { value } })
    const getValue = () => wrapper.find(TextField).props().value
    changeValue(5)
    await sleepAsync()
    expect(updateState).toHaveBeenCalled()
    expect(getValue()).toBe(5)

    // NaN
    changeValue('somestring')
    await sleepAsync()
    expect(getValue()).toBe(5)

    // limit min
    changeValue(-5)
    await sleepAsync()
    expect(getValue()).toBe(1)

    // limit max
    changeValue(100)
    await sleepAsync()
    expect(getValue()).toBe(10)
  })

  it('First page should jump to page 1', async () => {
    const wrapper = shallow(<PagerWidget />)
    wrapper.find('#FirstPage').simulate('click')
    await sleepAsync()
    expect(wrapper.find(TextField).props().value).toBe(1)
  })

  it('NavigateBefore should jump a page back', async () => {
    const wrapper = shallow(<PagerWidget />)
    wrapper.find('#NavigateBefore').simulate('click')
    await sleepAsync()
    expect(wrapper.find(TextField).props().value).toBe(4)
  })

  it('NavigateNext should jump to the next page', async () => {
    const wrapper = shallow(<PagerWidget />)
    wrapper.find('#NavigateNext').simulate('click')
    await sleepAsync()
    expect(wrapper.find(TextField).props().value).toBe(6)
  })

  it('Last page should jump to page 10', async () => {
    const wrapper = shallow(<PagerWidget />)
    wrapper.find('#LastPage').simulate('click')
    await sleepAsync()
    expect(wrapper.find(TextField).props().value).toBe(10)
  })
})
