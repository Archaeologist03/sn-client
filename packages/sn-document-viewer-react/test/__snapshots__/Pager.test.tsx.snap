// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagerWidget component Should render without crashing 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
    }
  }
>
  <WithStyles(IconButton)
    disabled={true}
    title="firstpage"
    onClick={[Function]}
  >
    <pure(FirstPageIcon)
    />
  </WithStyles(IconButton)>
  <WithStyles(IconButton)
    disabled={true}
    title="previouspage"
    onClick={[Function]}
  >
    <pure(NavigateBeforeIcon)
    />
  </WithStyles(IconButton)>
  <TextField
    InputLabelProps={
      Object {
        "shrink": true,
      }
    }
    inputProps={
      Object {
        "max": 10,
        "min": 1,
      }
    }
    margin="dense"
    onChange={[Function]}
    required={true}
    select={false}
    title="gotopage"
    type="number"
    value={1}
    variant="standard"
  />
  <WithStyles(IconButton)
    color="primary"
    disabled={false}
    title="nextpage"
    onClick={[Function]}
  >
    <pure(NavigateNextIcon)
    />
  </WithStyles(IconButton)>
  <WithStyles(IconButton)
    disabled={false}
    title="lastpage"
    onClick={[Function]}
  >
    <pure(LastPageIcon)
    />
  </WithStyles(IconButton)>
</div>
`;
